File: sdk/examples/how_tos/account_wallet/request_funds.rs
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we use an account as wallet.
//! Rename `.env.example` to `.env` first.
//!
//! `cargo run --release --all-features --example account_wallet_request_funds`

use iota_sdk::{
    client::request_funds_from_faucet,
    types::block::address::{AccountAddress, ToBech32Ext},
    wallet::{AccountSyncOptions, Result, SyncOptions},
    Wallet,
};

#[tokio::main]
async fn main() -> Result<()> {
    // This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    for var in ["FAUCET_URL", "WALLET_DB_PATH"] {
        std::env::var(var).expect(&format!(".env variable '{var}' is undefined, see .env.example"));
    }

    let faucet_url = std::env::var("FAUCET_URL").unwrap();

    // Create the wallet
    let wallet = Wallet::builder()
        .with_storage_path(&std::env::var("WALLET_DB_PATH").unwrap())
        .finish()
        .await?;

    let balance = wallet.sync(None).await?;

    let total_base_token_balance = balance.base_coin().total();
    println!("Balance before requesting funds on wallet address: {total_base_token_balance:#?}");

    let account_id = balance.accounts().first().unwrap();
    println!("Account Id: {account_id}");

    // Get account address
    let account_address = AccountAddress::new(*account_id).to_bech32(wallet.client().get_bech32_hrp().await.unwrap());
    let faucet_response = request_funds_from_faucet(&faucet_url, &account_address).await?;

    println!("{faucet_response}");

    tokio::time::sleep(std::time::Duration::from_secs(10)).await;

    let sync_options = SyncOptions {
        account: AccountSyncOptions {
            basic_outputs: true,
            ..Default::default()
        },
        ..Default::default()
    };
    let total_base_token_balance = wallet.sync(Some(sync_options)).await?.base_coin().total();
    println!("Balance after requesting funds on account address: {total_base_token_balance:#?}");

    Ok(())
}

File: bindings/python/examples/how_tos/account_output/request_funds.py
import os
import time

from dotenv import load_dotenv
from iota_sdk import AccountAddress, Wallet, WalletOptions, Utils, SyncOptions, WalletSyncOptions

# In this example we request funds to the wallet's first account output
# address.

load_dotenv()

for env_var in ['FAUCET_URL', 'WALLET_DB_PATH', 'EXPLORER_URL']:
    if env_var not in os.environ:
        raise Exception(f".env {env_var} is undefined, see .env.example")

FAUCET_URL = os.environ.get(
    'FAUCET_URL', 'https://faucet.testnet.shimmer.network/api/enqueue')

wallet = Wallet(WalletOptions(storage_path=os.environ.get('WALLET_DB_PATH')))
balance = wallet.sync(None)

total_base_token_balance = balance.base_coin.total
print(
    f"Balance before requesting funds to the wallet's first account output address: {total_base_token_balance}")

account_id = balance.accounts[0]
print(f'Account Id: {account_id}')

# Get Account address
account_address = Utils.address_to_bech32(
    AccountAddress(account_id), wallet.get_client().get_bech32_hrp())
faucet_response = wallet.get_client().request_funds_from_faucet(
    FAUCET_URL, account_address)
print(faucet_response)

time.sleep(10)

sync_options = SyncOptions(wallet=WalletSyncOptions(basic_outputs=True))

total_base_token_balance = wallet.sync(sync_options).base_coin.total
print(
    f"Balance after requesting funds to the wallet's first account output address: {total_base_token_balance}")
